To and create.

AGaliciaMartinez opened the pull request:
This PR adds:
- DenseTensor class that wraps around a TensorFlow `Tensor`.
- `to` and `create` functions for the dispatcher.

AGaliciaMartinez committed:
Import qutip to use it when importing qutip_tensorflow.

Since qutip_tensorflow will always be used together with qutip, it makes
sense to provide same functions by default.

AGaliciaMartinez committed:
First version of TensorFlow data backend.

AGaliciaMartinez committed:
Added VariablesDense.

AGaliciaMartinez committed:
Added initialize_tensor argument to VariableDense

AGaliciaMartinez committed:
Added Value Error if shape == None.

AGaliciaMartinez committed:
I remover variable. Will focus first on Tensor.

AGaliciaMartinez committed:
Changed name of class

AGaliciaMartinez committed:
Changed how copy works.

AGaliciaMartinez committed:
Added test (copied from Qutip)

AGaliciaMartinez committed:
Included test for different tensor input dtypes.

AGaliciaMartinez committed:
Added conftest, necessary for tesnting.

AGaliciaMartinez committed:
Added random_tensor_dense.

AGaliciaMartinez committed:
Added test from tensor_dense

AGaliciaMartinez committed:
Added test and expections to catch wrong shapes.

AGaliciaMartinez committed:
sort imports.

AGaliciaMartinez committed:
Tested and implemented copy method.

AGaliciaMartinez committed:
Added test to ensure output is a copy

AGaliciaMartinez committed:
Cleaner version.

AGaliciaMartinez committed:
No view in tensorflow so no test needed.

AGaliciaMartinez committed:
Removed test related to as_ndarray

AGaliciaMartinez committed:
Implemented to_array

AGaliciaMartinez committed:
Small modification.

AGaliciaMartinez committed:
Added test for initialization with list.

AGaliciaMartinez committed:
Removed unnecessary methods.

AGaliciaMartinez committed:
Removed unnecessary functions in tests.

AGaliciaMartinez committed:
Added __init__ files.

AGaliciaMartinez committed:
__init__ now imports all from core.

AGaliciaMartinez committed:
Added test for to method.

AGaliciaMartinez committed:
Added to_method

AGaliciaMartinez committed:
Added docstring for the __init__ function.

AGaliciaMartinez committed:
PEP8

AGaliciaMartinez committed:
Renamed file.

AGaliciaMartinez committed:
Merge branch 'main' into tensor_implementation

AGaliciaMartinez committed:
Merge branch 'tensor_implementation' into to_method_dispatcher

AGaliciaMartinez committed:
Changed tensorflow alias to tftensor

AGaliciaMartinez committed:
Fixed tests and imports.

AGaliciaMartinez committed:
Merge branch 'tensor_implementation' into to_method_dispatcher

AGaliciaMartinez committed:
Fixed gub from changing name of file.

AGaliciaMartinez committed:
Added create for Tensor.

AGaliciaMartinez committed:
Changed copy to False so that create keeps the Tensor structure.

AGaliciaMartinez commented:
One of the things I had to do here is set copy=False in the __init__ of `DenseTensor`. This is so that `create` maintains the graph structure as I would say this should be the default behaviour. This is different to what happens with `Dense` where the default value of `copy` is `True`. Should I keep it false to maintain the graph structure or is there a way to keep the graph by default when using create but still having `copy=True` in `DenseTensor`?

AGaliciaMartinez committed:
Added test for create and fixed test for copy=False.

AGaliciaMartinez commented on src/qutip_tensorflow/core/data/convert.py:
```diff
@@ -0,0 +1,33 @@
+    (is_tftensor, DenseTensor, 85),
+])
+
+to = qt.data.to
+create = qt.data.create
```
Should I make to and create accessible from qtf?

jakelishman replied:
No - there's almost never a reason to re-expose the same object from a different location, particularly not across library boundaries.  The point of a dispatcher plugin is that the data-types should work from the main QuTiP library for basic use-cases, otherwise you'll cause confusion - if a user mutates the `qutip_tensorflow.to` object (e.g. by adding aliases or changing the preferred conversion rules), it would be a huge surprise to discover that `qutip.to` has also changed.

AGaliciaMartinez committed:
Fixed bug due to change in name.

AGaliciaMartinez committed:
Merge branch 'tensor_implementation' into to_method_dispatcher

jakelishman commented on src/qutip_tensorflow/core/data/convert.py:
```diff
@@ -0,0 +1,33 @@
+import qutip as qt
```
I would argue there's no reason to abbreviate `qutip`.  It's short already, and it just hurts readability.

jakelishman commented on src/qutip_tensorflow/core/data/convert.py:
```diff
@@ -0,0 +1,33 @@
+])
+
+# You can add user friendly name for conversion with `to` or Qobj creation functions:
+qt.data.to.register_aliases(['tftensor'], DenseTensor)
+qt.data.to.register_aliases(['DenseTensor'], DenseTensor)
```
the first argument is a list, specifically so you can pass more than one thing in a single call

jakelishman commented on src/qutip_tensorflow/core/data/dense_tensor.py:
```diff
@@ -0,0 +1,85 @@
+import qutip as qt
```
again, really not sure I see a reason to abbreviate this.  People read code more than they write it.

jakelishman commented on src/qutip_tensorflow/core/data/dense_tensor.py:
```diff
@@ -0,0 +1,85 @@
+import qutip as qt
+import numpy as np
+import tensorflow as tf
+import qutip.core.data as data
```
You should protect this name with an underscore or something - you frequently use `data` as a variable, so at several points during the code, it's not at all clear what the identifier `data` refers to.

jakelishman commented on src/qutip_tensorflow/core/data/dense_tensor.py:
```diff
@@ -0,0 +1,85 @@
+        it needs to be moved to the GPU.
+    """
+
+    def __init__(self, data, shape=None, copy=False):
+        # Try to inherit shape from data
+        if shape is None:
+            try:
+                shape = data.shape
+            except AttributeError:
+                raise ValueError("""Shape could not be inferred from data.
+                                 Please, include the shape of data.""")
```
It's not clear to me what inputs you're prepared to accept.  The docstrings says `array_like`, but this check will cause `DenseTensor([[1, 0], [0, 1]])` to raise an exception, despite it being a valid Numpy array-like object.  You might want to rethink the order you do things, so you're not repeating checks or throwing away valid input from which the shape _can_ be inferred.  Remember that the `__init__` method is really meant for _humans_ more than anything else - library code often uses fast alternative constructor methods, because it often already has a lot of knowledge about the data being passed in, and having to write more doesn't matter very much in library code.

jakelishman commented on src/qutip_tensorflow/core/data/dense_tensor.py:
```diff
@@ -0,0 +1,85 @@
+
+        super().__init__(shape)
+
+        try:
+            self._tf = tf.constant(data, shape=shape)
+        except TypeError as e:
+            raise ValueError("Shape of data must match shape argument.") from e
```
This error handling doesn't look right.  What if `data` is some random object which has a `shape` attribute, but isn't array-like?  Then the error is that invalid data has been passed, but `tf.constant(data)` will still likely throw a TypeError, so you'll see this error.

AGaliciaMartinez replied:
I changed code logic. I think now this issue would be solved.

jakelishman commented on tests/core/data/test_tensordense.py:
```diff
@@ -0,0 +1,234 @@
+        with pytest.raises(error):
+            DenseTensor(arg, **kwargs)
+
+    def test_copy_returns_a_correct_copy(self, data_tensor_dense):
+        """
+        Test that the copy() method produces an actual copy, and that the
+        result represents the same matrix.
+        """
+        original = data_tensor_dense
+        copy = data_tensor_dense.copy()
+        assert original is not copy
+        assert np.all(original._tf == copy._tf)
```
You should also check that mutations of the original don't affect the copy, and vice-versa.

AGaliciaMartinez replied:
What do you mean by mutations? Could you provide me an example with when this could go wrong?  

If by mutations you mean that I should check that an "in place" modification of original._tf does no change copy._tf (for example by doing something like: `original._tf[0, 0] += 1`) then, it turns out that for `Tensor` it is not possible to modify it in place. For instance, `original._tf[0, 0] += 1` will return `TypeError`. `Variables` from TensorFlow can but I plan to implement them in another class.

In any case, thinking about your comment I realised that this can happen:
```python
original = data_tensor_dense
maybe_copy = DenseTensor(original._tf, copy=False) # This is not what happens with copy() method but could happen!
print(original is not maybe_copy) # True
print(np.all(original._tf == maybe_copy._tf)) #True
print(maybe_copy._tf is not original._tf) # False!!!!!!
```
So I will add in the tests: `assert copy._tf is not original._tf`. In any case I am not sure if this would matter that much as tensors can not be modified "in place".

jakelishman commented on tests/core/data/test_tensordense.py:
```diff
@@ -0,0 +1,234 @@
+                                       'int32', 'int64',
+                                       'uint32'])
+    def test_init_from_list_other_dtype(self, shape, dtype):
+        numpy_dense = np.random.rand(*shape).astype(dtype, casting='unsafe')
+        list_dense = numpy_dense.tolist()
```
You're shadowing a fixture name here, which is unclear - it means the same name refers to quite different things through the code.

AGaliciaMartinez replied:
Thanks for catching this. I named them _list_dense _numpy_dense. I would suggest doing the same in QuTiP tests as numpy_dense is sometimes used as a fixture and some other times not. ( I actually thought it did not matter because QuTiP tests already did it). I can make a pull request if in QuTiP if you agree.

jakelishman commented on tests/core/data/test_tensordense.py:
```diff
@@ -0,0 +1,234 @@
+        assert test.shape == shape
+        assert test._tf.dtype == tf.complex128
+        assert test._tf.shape == shape
+        assert_almost_equal(test.to_array(), list_dense)
```
Are they really allowed to only be _almost_ equal?

AGaliciaMartinez replied:
With `assert np.all(test.to_array()==np.array(_list_dense))` it fails for:

FAILED tests/core/data/test_tensordense.py::TestClassMethods::test_init_from_list_other_dtype[ket-float64] - assert False
FAILED tests/core/data/test_tensordense.py::TestClassMethods::test_init_from_list_other_dtype[bra-float64] - assert False
FAILED tests/core/data/test_tensordense.py::TestClassMethods::test_init_from_list_other_dtype[square-float64] - assert False
FAILED tests/core/data/test_tensordense.py::TestClassMethods::test_init_from_list_other_dtype[wide-float64] - assert False
FAILED tests/core/data/test_tensordense.py::TestClassMethods::test_init_from_list_other_dtype[tall-float64] - assert False


Could this be due to the casting rules? I looks weird as it only fails for float64.

jakelishman commented on tests/core/data/test_tensordense.py:
```diff
@@ -0,0 +1,234 @@
+        Test that the __init__ method raises a suitable error when passed
+        incorrectly formatted inputs.
+
+        This test also serves as a *partial* check that Dense safely handles
+        deallocation in the presence of exceptions in its __init__ method.  If
+        the tests segfault, it's quite likely that the memory management isn't
+        being done correctly in the hand-off us setting our data buffers up and
+        marking the numpy actually owns the data.
```
Please don't just copy-and-paste code - incorrect documentation can even be worse than no documentation at all.

AGaliciaMartinez replied:
Sorry for this. I will go through all the comments in the test and ensure they are correct.

AGaliciaMartinez committed:
Added PR changes.

AGaliciaMartinez committed:
Added PR changes.

AGaliciaMartinez committed:
Change logic to behave properly (with lists) and be more clear of what is happening.

AGaliciaMartinez committed:
Small change to docstring.

AGaliciaMartinez committed:
Icluded some comments for clarity.

AGaliciaMartinez committed:
Changed test to not give a shape argument to DenseTesor when input is a list.

In previous versions of DenseTensor this was overlooked and list input
only worked if shape was also provided.

AGaliciaMartinez committed:
Added test to ensure that _tf is a copy after using the method copy.

AGaliciaMartinez committed:
Changed list_dense and numpy_dense names when they are not fixtures.

I simply preceeded them with an underscore.

AGaliciaMartinez committed:
I forgot one in previous commit.

AGaliciaMartinez committed:
One more that I forgot.

AGaliciaMartinez committed:
Fixed Bug: wrong error message for when shape is an integer.

AGaliciaMartinez committed:
Simpliefied tests for wrong shape input.

AGaliciaMartinez committed:
Checked that docstrings are correct.

AGaliciaMartinez committed:
PEP 8 with black.

AGaliciaMartinez committed:
Remove unused import.

AGaliciaMartinez commented on tests/core/data/test_tensordense.py:
```diff
@@ -0,0 +1,170 @@
+    return request.param
+
+
+@pytest.fixture(params=[True, False], ids=["Fortran", "C"])
+def fortran(request):
+    return request.param
```
Remove this

AGaliciaMartinez replied:
Actually, do not remove this as `numpy_dense` creates an array that depends on this fixture. For example, it necessary to test that `DenseTensor` is created properly from a Fortran array. Although, I am not sure how important is this.

AGaliciaMartinez commented on tests/core/data/test_tensordense.py:
```diff
@@ -0,0 +1,170 @@
+            pytest.param((1, 2, 3), id="shape_3_tuple"),
+            pytest.param((-1, 1), id="negative_shape"),
+            pytest.param((-4, -4), id="both_negative_shape"),
+            pytest.param((1213, 1217), id="differen_shape"),
```
check what happens if shape input not the same as data input

AGaliciaMartinez committed:
Added test with correct shape but different values from data.

AGaliciaMartinez committed:
Renamed DenseTensor to TfTensor.

AGaliciaMartinez commented:
I added the missing test and renamed DenseTensor to TfTensor. I would like to merge this PR by tomorrow (@Ericgig agrees). @jakelishman, would you like to suggest any other change or are you happy with the merge?

jakelishman commented on src/qutip_tensorflow/core/data/tftensor.py:
```diff
@@ -0,0 +1,98 @@
+        `Tenor`) or if a copy is needed to satisfy dtype (tensorflow.complex128) and
+        shape (2D Tensor)."""
+
+    def __init__(self, data, shape=None, copy=False):
```
Is there any situation in which the backing data field of this object can be mutable?  I don't know tensorflow so well, but say for numpy
```python
>>> class Data:
...   def __init__(self, data):
...      self.data = np.array(data, copy=False)
...   def __repr__(self):
...     return f"Data({repr(self.data)})"
...
>>> x = np.ones((2,))
>>> y = Data(x)
>>> x
array([1., 1.])
>>> y
Data(array([1., 1.]))
>>> x *= 2
>>> x
array([2., 2.])
>>> y
Data(array([2., 2.]))
```
so the derived object has changed from under us.

If there is such a case, then the default argument to `copy` _must_ be `True`.  If not, then does the `copy` keyword argument serve any purpose?

AGaliciaMartinez replied:
The above case would not work for tensorflow if using a `Tensor`:
```python
a = random_tensor((1,1)) # random tensor of shape (1,1)
b = TfTensor(a)
b._tf == a # True
a *= 2
b._tf == a #False
```
To my knowledge, there is no operation in a `tf.Tensor` that does not return a different tensor. `tf.Variables`, although not supported yet in qutip-tensorflow, do perform in-place operations with the `assign()` method. 

I can not think of a case where `copy=true` is necessary but it does have a different behaviour. For instance,
```python
a = tf.constant([[1+1j]])
b = TfTensor(a, copy=False) 
a is not b._tf # False
b = TfTensor(a, copy=True)
a is not b._tf # True 
``` 
Although I do not know when and why this could be useful (this may be why `tf.constant()` does not have a `copy` argument).

jakelishman replied:
If tensors are immutable, then the copy argument is pretty redundant. It's fine to have it for consistency, you may just want to mention that it does very little in the docstring.

jakelishman commented on src/qutip_tensorflow/core/data/tftensor.py:
```diff
@@ -0,0 +1,98 @@
+
+class TfTensor(qutip.core.data.Data):
+    """This class provide a wraps around TensorFlow's Tensor. It will store data
+    as a Tensor of dtype `tensorflow.complex128`. Data will be expanded into a 2D
```
you need double back-ticks (everywhere) to make this render as inline code, as opposed to in italics.

jakelishman commented on src/qutip_tensorflow/core/data/tftensor.py:
```diff
@@ -0,0 +1,98 @@
+
+        # Try to inherit shape from data and expand shape
+        if shape is None:
+            try:
+                shape = data.shape
+            except AttributeError:
+                raise ValueError(
+                    """Shape could not be inferred from data.
+                                 Please, include the shape of data."""
+                )
```
`data` is now guaranteed to be a tensor, so is it ever possible for this to raise an exception?

AGaliciaMartinez replied:
You are right, this was necessary for the previous logic but is not any more. Removing.

jakelishman commented on src/qutip_tensorflow/core/data/tftensor.py:
```diff
@@ -0,0 +1,98 @@
+                    """Shape could not be inferred from data.
+                                 Please, include the shape of data."""
+                )
+
+            if isinstance(shape, tf.TensorShape):
+                shape = tuple(shape.as_list())
```
Similarly, is it ever possible for this _not_ to be the case?

jakelishman commented on src/qutip_tensorflow/__init__.py:
```diff
@@ -1 +1,2 @@
 from .version import version as __version__
+from .core import *
```
Do you need a separate `core` package within `qutip_tensorflow`?  I wouldn't imagine there'll be too many different levels in it, so this may just be adding needless structure.

AGaliciaMartinez replied:
I was following the same structure that QuTiP has for consistency. It is no necessary but I though someone working on QuTiP would find it more appealing.

jakelishman replied:
I'd say there's no need to try and maintain structural consistency between separate packages. QuTiP's core is just for code organisation anyway - it's not directly exposed to users.

jakelishman commented on tests/core/data/test_convert.py:
```diff
@@ -0,0 +1,80 @@
+    ],
+)
+def test_converters_qt_to_qtf(from_, base, to_, dtype):
+    print(base)
```
this `print` doesn't serve a purpose - pytest eats the stdout stream on success, and would print the parameters to the test on failure.

jakelishman commented on tests/core/data/test_convert.py:
```diff
@@ -0,0 +1,80 @@
+        pytest.param(TfTensor, TfTensor, id="to tensorflow type"),
+    ],
+)
+def test_converters_qtf_to_qt(from_, base, to_, dtype):
+    converter = data.to[to_, from_]
+    assert isinstance(converter(base), dtype)
+    converter = data.to[to_]
+    assert isinstance(converter(base), dtype)
+    assert isinstance(data.to(to_, base), dtype)
```
I'm not sure what exactly the test (and the similar reverse test) is meant for.  The behaviour seems to be a lot more of a test of `qutip.data.to` than anything else - it seems to just be testing that the alias and partial application systems work.  If your intent is to test that the various conversions have truly been added, you probably want to investigate the `direct` attribute of the output of `data.to[to_, from_]`.

AGaliciaMartinez replied:
This is quite a change and I really do not want to delay this any more. I will merge this pull request so that I can open the next. I will also open a new PR making this test more sensible.

jakelishman commented on tests/core/data/test_tftensor.py:
```diff
@@ -0,0 +1,171 @@
+        test = TfTensor(tensor_dense)
+        assert test.shape == tuple(tensor_dense.shape.as_list())
+        assert np.all(test.to_array() == tensor_dense)
+
+        # by default we do not return a copy
+        assert test._tf is tensor_dense
```
If so, you probably also want to check that using the `copy` keyword _does_ return a copy.

jakelishman commented on tests/core/data/test_tftensor.py:
```diff
@@ -0,0 +1,171 @@
+        assert_almost_equal(test.to_array(), _list_dense)
+
+    @pytest.mark.parametrize(
+        "dtype", ["complex128", "float64", "int32", "int64", "uint32"]
+    )
+    def test_init_from_ndarray_other_dtype(self, shape, dtype):
+        _numpy_dense = np.random.rand(*shape).astype(dtype, casting="unsafe")
```
Be careful with this casting; `np.random.rand()` returns a float on [0, 1], which will be cast to a integer 0 every time bar ~once every 2^53 times.  So you're basically just testing on the zero matrix every time, which probably isn't what you meant.

AGaliciaMartinez replied:
I see the problem but given there is such a low chance to occur I think I will let it be for the moment and focus on the next PR.

jakelishman replied:
The point is more that the test isn't correct - you're not actually passing it random data, you're just passing a zero matrix. You could basically just fix it by generating random unsigned integers instead of random floats, and then you wouldn't need the unsafe cast either.

jakelishman reviewed (commented):
I have some comments, a few of which may have been addressed in meetings that I've not been in (sorry).  Mostly it looks fine, though, with Eric's comment too, provided Eric's also happy with the default `copy=False` in the constructor.

AGaliciaMartinez committed:
Removed unnecessary logic in `TfTensor`

AGaliciaMartinez committed:
Changed docstring for copy.

AGaliciaMartinez committed:
Changed single back-ticks to double back-ticks.

AGaliciaMartinez committed:
Removed unnecessary print.

AGaliciaMartinez committed:
Added test for copy argument in TfTensor.

AGaliciaMartinez committed:
Test for ndarray input.

AGaliciaMartinez commented:
Thanks @jakelishman and @Ericgig for reviewing the PR! I will merge this now and fix some minor comments in another PR. This way the next PR will be ready for review today.
